### PR TITLE
DatabaseTarget - Added support for Connection Properties (e.g. for Azure AD Access Token)

### LINKED ISSUE #3554: Configure and write to Azure SQL server with manage Identity token
I want to connect with Managed Identity and write the logs in Azure SQL. How can I configure this. As manage identity login has an Access token.

**NLog version**: 4.6.6

**Platform**: . .NET Core 2

**Current NLog config** (4.6.6)

 - Why do we need it?
 - Need to log error with Managed Identity in Azure SQL

### PR DESCRIPTION
**Pull request by snakefoot**
Based on  #3825

fixes #3554

```xml
<target name="db"
        xsi:type="Database"
        commandType="StoredProcedure"
        commandText="[dbo].[NLog_AddEntry_p]"
        >
  <connectionProperty name="AccessToken" layout="${gdc:AccessToken}" propertyType="System.String" />
  <parameter name="@logged"         layout="${date}" />
  <parameter name="@level"          layout="${level}" />
  <parameter name="@message"        layout="${message}" />
  <parameter name="@logger"         layout="${logger}" />
</target>
```

For SqlConnection AccessToken then one needs a special LayoutRenderer that can call `AzureServiceTokenProvider` as AccessToken can have a limited lifetime.

### CODE REVIEW

**Comment by 304NotModified** [comment]
Could you please add a config example for this great feature for all future readers?

File: src/NLog/Internal/PropertySetter.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+namespace NLog.Internal
+{
+    using System;
+    using System.Reflection;
+    using NLog.Common;
+
+    internal interface IPropertySetter
```
Comment (304NotModified) [inline src/NLog/Internal/PropertySetter.cs]: Please move to a separate file

File: src/NLog/Internal/PropertySetter.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        bool SetPropertyValue(object instance, object value);
+    }
+
+    internal class PropertySetter : IPropertySetter
+    {
+        private readonly Type _objectType;
+        private readonly PropertyInfo _propertyInfo;
+        private ReflectionHelpers.LateBoundMethod _propertySetter;
+
+        public static IPropertySetter GeneratePropertySetter(Type objectType, string propertyName)
```
Comment (304NotModified) [inline src/NLog/Internal/PropertySetter.cs]: Please CreatePropertySetter (it wont generate things IMO)

File: src/NLog/Internal/PropertySetter.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+            if (_propertySetter != null)
+            {
+                _propertySetter.Invoke(instance, new[] { value });
+            }
+            else
+            {
+                // Generate compiled method if setter works without throwing exception
+                _propertyInfo.GetSetMethod().Invoke(instance, new[] { value });
+                _propertySetter = ReflectionHelpers.CreateLateBoundMethod(_propertyInfo.GetSetMethod());
```
Comment (304NotModified) [inline src/NLog/Internal/PropertySetter.cs]: Maybe doing once `_propertyInfo.GetSetMethod()`? (sounds like an expensive method).

Maybe describe this class as "lazy compiled" (In class XML docs)

**Comment by snakefoot** [comment]
@anxkha Looks like it is your lucky day. Getting the feature made for you almost on a silver plate. Links to nuget-package here:

https://ci.appveyor.com/project/nlog/nlog/builds/31250180/artifacts

Have added a new collection on DatabaseTarget called `ConnectionProperties` there you can add custom properties to assign on the SqlConnection.

```c#
var accessTokenLayout = NLog.Layouts.Layout.FromMethod(logEvent => "MyToken");
databaseTarget.ConnectionProperties.Add(new DatabaseConnectionInfo() { Name = "AccessToken", Layout = accessTokenLayout });
```

Hope it lives up to your expectations.

File: src/NLog/Targets/DatabaseConnectionInfo.cs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        /// <docgen category='Connection Options' order='10' />
+        [RequiredParameter]
+        public string Name { get; set; }
+
+        /// <summary>
+        /// Gets or sets the value to assign on DbConnection
+        /// </summary>
+        /// <docgen category='Connection Options' order='10' />
+        [RequiredParameter]
+        public Layout Layout { get; set; }
```
Comment (304NotModified) [inline src/NLog/Targets/DatabaseConnectionInfo.cs]: Could we call this Value?

File: src/NLog/Targets/DatabaseConnectionInfo.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DatabaseConnectionInfo"/> class.
+        /// </summary>
+        public DatabaseConnectionInfo()
+        {
+        }
+
+        /// <summary>
+        /// Gets or sets the DbConnection Property to update
```
Comment (304NotModified) [inline src/NLog/Targets/DatabaseConnectionInfo.cs]: please mention it's not case sensitive

File: src/NLog/Targets/DatabaseConnectionInfo.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+        /// <summary>
+        /// Gets or sets the type of the property
+        /// </summary>
+        /// <docgen category='Connection Options' order='10' />
+        [DefaultValue(typeof(string))]
+        public Type PropertyType { get; set; } = typeof(string);
+
+        /// <summary>
+        /// Gets or sets convert format of the property value .
```
Comment (304NotModified) [inline src/NLog/Targets/DatabaseConnectionInfo.cs]: floating dot ;)

Comment (304NotModified) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: Please describe when we use this format in the XML docs.

File: src/NLog/Targets/DatabaseConnectionInfo.cs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    using System.Globalization;
+    using NLog.Config;
+    using NLog.Internal;
+    using NLog.Layouts;
+
+    /// <summary>
+    /// Information about database command + parameters.
+    /// </summary>
+    [NLogConfigurationItem]
+    public class DatabaseConnectionInfo
```
Comment (304NotModified) [inline src/NLog/Targets/DatabaseConnectionInfo.cs]: Maybe name it DatabaseConnectionProperty?

File: src/NLog/Targets/DatabaseConnectionInfo.cs [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    using System.Collections.Generic;
+    using System.ComponentModel;
+    using System.Data;
+    using System.Globalization;
+    using NLog.Config;
+    using NLog.Internal;
+    using NLog.Layouts;
+
+    /// <summary>
+    /// Information about database command + parameters.
```
Comment (304NotModified) [inline src/NLog/Targets/DatabaseConnectionInfo.cs]: Wrong summary?

File: src/NLog/Targets/DatabaseTarget.cs [hunk 9]
```diff
@@ -657,7 +692,7 @@ private void WriteLogEventBatchToDatabase(IList<AsyncLogEventInfo> logEvents, st
                 //Always suppress transaction so that the caller does not rollback logging if they are rolling back their transaction.
                 using (TransactionScope transactionScope = new TransactionScope(TransactionScopeOption.Suppress))
                 {
-                    EnsureConnectionOpen(connectionString);
+                    EnsureConnectionOpen(connectionString, logEvents.Count > 0 ? logEvents[0].LogEvent : null);
```
Comment (304NotModified) [inline src/NLog/Targets/DatabaseTarget.cs]: There was a reason for not using Linq, isn't?

File: src/NLog/Targets/DatabaseConnectionInfo.cs [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        public CultureInfo Culture { get; set; }
+
+        internal bool SetPropertyValue(IDbConnection dbConnection, object propertyValue)
+        {
+            var dbConnectionType = dbConnection.GetType();
+            var propertySetterCache = _propertySetter;
+            if (propertySetterCache.Key.Key != Name || propertySetterCache.Key.Value != dbConnectionType)
+            {
+                var propertySetter = PropertySetter.GeneratePropertySetter(dbConnectionType, Name);
+                propertySetterCache = new KeyValuePair<KeyValuePair<string, Type>, IPropertySetter>(new KeyValuePair<string, Type>(Name, dbConnectionType), propertySetter);
```
Comment (304NotModified) [inline src/NLog/Targets/DatabaseConnectionInfo.cs]: I really don't like to KeyValuePair (unreadable)... as we can't use c# 7 tuples, I propose:

- Creating a new type (class or struct) for ` KeyValuePair<KeyValuePair<string, Type>, IPropertySetter>`
- could be private nested
- Implement Equals for the check below.

e.g.

```c#
private class PropertySetterCacheItem
{
    [NotNull] public string Name { get; }
    [NotNull] public Type Type { get; }
    [NotNull] public IPropertySetter PropertySetter { get; }

    public PropertySetterCacheItem([NotNull] string name, [NotNull] Type type, [NotNull] IPropertySetter propertySetter)
    {
        Name = name ?? throw new ArgumentNullException(nameof(name));
        Type = type ?? throw new ArgumentNullException(nameof(type));
        PropertySetter = propertySetter ?? throw new ArgumentNullException(nameof(propertySetter));
    }


    /// <summary>
    /// Equals without IPropertySetter
    /// </summary>
    protected bool Equals(string name, Type type)
    {
        return string.Equals(Name, name, StringComparison.OrdinalIgnoreCase) && Type == type;
    }


}
```

File: src/NLog/Targets/DatabaseTarget.cs [hunk 11]
```diff
@@ -1043,77 +1078,68 @@ protected virtual IDbDataParameter CreateDatabaseParameter(IDbCommand command, D
...
-            {
-                return RenderLogEvent(parameterInfo.Layout, logEvent) ?? string.Empty;
-            }
+            return RenderObjectValue(logEvent, parameterInfo.Name, parameterInfo.Layout, parameterInfo.ParameterType, parameterInfo.Format, parameterInfo.Culture);
+        }
 
-            IFormatProvider dbParameterCulture = GetDbParameterCulture(logEvent, parameterInfo);
+        private object GetDatabaseConnectionValue(LogEventInfo logEvent, DatabaseConnectionInfo connectionInfo)
+        {
+            return RenderObjectValue(logEvent, connectionInfo.Name, connectionInfo.Layout, connectionInfo.PropertyType, connectionInfo.Format, connectionInfo.Culture);
```
Comment (304NotModified) [inline src/NLog/Targets/DatabaseTarget.cs]: proposal: create an interface for DatabaseParameterInfo and DatabaseConnectionProperty, and pass that interface to RenderObjectValue.

**Comment by snakefoot** [comment]
Guess it could evolve into also having the ability to assign "CommandProperty" :)

File: src/NLog/Targets/DatabaseTarget.cs (hunk 11 above)
Comment (304NotModified) [reply src/NLog/Targets/DatabaseTarget.cs]: We need to add some internal readonly properties for the common interface on DatabaseParameterInfo and DatabaseConnectionProperty then, e.g.ValueType and ValueLayout

Comment (304NotModified) [reply src/NLog/Targets/DatabaseTarget.cs]: e.g. 


```c#
internal interface IDatabaseRenderable
{
    string Name { get; }
    Layout Layout { get; }
    Type ValueType { get; }
    string Format { get; }
    CultureInfo Culture { get; }
}

```

File: src/NLog/Targets/DatabaseConnectionInfo.cs (hunk 4 above)
Comment (snakefoot) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: Nope Layout is the official standard in NLog.

File: src/NLog/Targets/DatabaseConnectionInfo.cs (hunk 6 above)
Comment (snakefoot) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: Exotic scenario that is hard to explain.

File: src/NLog/Targets/DatabaseConnectionInfo.cs (hunk 7 above)
Comment (snakefoot) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: I want it to match DatabaseParameterInfo

File: src/NLog/Targets/DatabaseConnectionInfo.cs (hunk 8 above)
Comment (snakefoot) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: Lots of copy-paste

File: src/NLog/Targets/DatabaseTarget.cs (hunk 9 above)
Comment (snakefoot) [reply src/NLog/Targets/DatabaseTarget.cs]: So many traps in Linq, so I try to avoid it, unless writing unit-testing code. Then allocation is free.

File: src/NLog/Targets/DatabaseConnectionInfo.cs (hunk 10 above)
Comment (snakefoot) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: Sure just wanted to make it quick, since you were so eager on making a half solution. Now suddenly it has changed :)

File: src/NLog/Targets/DatabaseTarget.cs (hunk 11 above)
Comment (snakefoot) [reply src/NLog/Targets/DatabaseTarget.cs]: Sounds like an mental exercise for you.

File: src/NLog/Targets/DatabaseConnectionInfo.cs (hunk 7 above)
Comment (304NotModified) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: - multiple DatabaseParameterInfo means multiple parameters
- multiple DatabaseConnectionInfo doesn't mean multiple connections ;)

So the same name is a bit confusing

File: src/NLog/Targets/DatabaseTarget.cs (hunk 9 above)
Comment (304NotModified) [reply src/NLog/Targets/DatabaseTarget.cs]: Could we add our own .FirstOrDefault (now or in the future)? It;s some much cleaner

File: src/NLog/Internal/PropertySetter.cs (hunk 3 above)
Comment (snakefoot) [reply src/NLog/Internal/PropertySetter.cs]: Not expensive, but I can move it into a local variable for you.

File: src/NLog/Internal/PropertySetter.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog/Internal/PropertySetter.cs]: Ok

File: src/NLog/Internal/PropertySetter.cs (hunk 3 above)
Comment (304NotModified) [reply src/NLog/Internal/PropertySetter.cs]: I like locals :)

File: src/NLog/Internal/PropertySetter.cs (hunk 1 above)
Comment (snakefoot) [reply src/NLog/Internal/PropertySetter.cs]: Removed the interface

File: src/NLog/Targets/DatabaseConnectionInfo.cs (hunk 5 above)
Comment (snakefoot) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: Ok

File: src/NLog/Targets/DatabaseConnectionInfo.cs (hunk 10 above)
Comment (304NotModified) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: LOL I had the same with #3825 ...

File: src/NLog/Targets/DatabaseTarget.cs (hunk 11 above)
Comment (304NotModified) [reply src/NLog/Targets/DatabaseTarget.cs]: Yeah if you OK with this, I could push it to this branch

**Comment by 304NotModified** [comment]
> For SqlConnection AccessToken then one needs a special LayoutRenderer that can call `AzureServiceTokenProvider` as AccessToken can have a limited lifetime.

So after all, an AccessToken property is maybe easier to use?

File: src/NLog/Targets/DatabaseConnectionInfo.cs (hunk 10 above)
Comment (304NotModified) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: Copy paste my private class is also quick? ;)

Comment (snakefoot) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: But now changed into struct.

Comment (304NotModified) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: OK, I was under the impression that a ref type in a struct, makes a struct a bit useless.

**Comment by snakefoot** [comment]
> So after all, an AccessToken property is maybe easier to use?

Not sure I understand the question

**Comment by 304NotModified** [comment]
Well I like the generic/reusable approach. But it it will be a lot harder to use for an user, then i'm not sure if shouldn't just add the DbAccessToken property (and do the magic in there)

File: src/NLog/Targets/DatabaseConnectionInfo.cs (hunk 7 above)
Comment (snakefoot) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: Not sure I understand. DatabaseConnectionInfo resides in a list. So does DatabaseParameterInfo

File: src/NLog/Targets/DatabaseTarget.cs (hunk 9 above)
Comment (snakefoot) [reply src/NLog/Targets/DatabaseTarget.cs]: Feel free

**Comment by snakefoot** [comment]
> add the DbAccessToken property (and do the magic in there)

What kind of magic will you do to wield AzureServiceTokenProvider without any dependencies ? Actually the reason why I wanted  @anxkha to perform this whole exercise, since the actual user. Never used an AccessToken in my life (Would allow us to learn something about how it used in different scenarios. AzureFunctions. AzureAspNetCore, etc. along with different lifetime-types).

File: src/NLog/Targets/DatabaseConnectionInfo.cs (hunk 10 above)
Comment (snakefoot) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: I like immutable structs. They are clean and doesn't allocate. And provide nearness in memory.

File: src/NLog/Targets/DatabaseTarget.cs (hunk 11 above)
Comment (snakefoot) [reply src/NLog/Targets/DatabaseTarget.cs]: Sure push away. Have committed the code-review changes agreed upon.

**Comment by snakefoot** [comment]
One extra push was needed. Refactoring introduced a bug.

File: src/NLog/Targets/DatabaseTarget.cs (hunk 11 above)
Comment (304NotModified) [reply src/NLog/Targets/DatabaseTarget.cs]: OK unresolve as reminder

**Comment by 304NotModified** [comment]
Yes I was already typing a comment and the commit was magically disappeared :P

File: src/NLog/Targets/DatabaseConnectionInfo.cs (hunk 7 above)
Comment (304NotModified) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: - **Two** DatabaseParameterInfos will create/work on **two** database parameters. 
- **Two** DatabaseConnectionInfos work on **one single** DBconnection.

The same name won't work well if the don't behave the same. 

Dunno how I could explain this differently.

File: src/NLog/Targets/DatabaseConnectionInfo.cs (hunk 4 above)
Comment (304NotModified) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: More like semiofficial ;)

File: src/NLog/Targets/DatabaseConnectionInfo.cs (hunk 7 above)
Comment (snakefoot) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: Let me rephrase your sentence:

- **Two** DatabaseParameterInfos will assign two parameters to a DbCommand
- **Two** DatabaseConnectionInfos will assign two properties to a DbConnection

And both names makes sense to me.

Comment (304NotModified) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: Maybe it's DatabaseConnectionPropertyInfo then

Comment (304NotModified) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: > Let me rephrase your sentence:
> 
> * **Two** DatabaseParameterInfos will assign two parameters to a DbCommand 
> * **Two** DatabaseConnectionInfos will assign two properties to a DbConnection
> 

Don't you see the inconsistency? ParameterInfos/DbCommand  vs ConnectionInfos/DbConnection?

Any way, consistency aside, I find the name  ConnectionInfo confusing

Comment (snakefoot) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: Ok lets rename it to DatabasePropertyInfo. Then it can also be reused for DbCommand-properties.

Comment (304NotModified) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: Also the same in the target code:

c#
```
  public IList<DatabaseParameterInfo> Parameters { get; } = new List<DatabaseParameterInfo>();
 public IList<DatabaseConnectionInfo> ConnectionProperties { get; } = new List<DatabaseConnectionInfo>();
```

- **Parameter**s  - Database**Parameter**Info
- **ConnectionProperties**  - DatabaseConnectionInfo

**Comment by anxkha** [comment]
@snakefoot @304NotModified Well, thank you both for taking this on. I've been super busy the last few days and you both moved so fast on this I didn't even have a good opportunity to sit down and fully review everything done in the PRs and your discussions.

> var accessTokenLayout = NLog.Layouts.Layout.FromMethod(logEvent => "MyToken");
> databaseTarget.ConnectionProperties.Add(new DatabaseConnectionInfo() { Name = "AccessToken", Layout = accessTokenLayout });

I haven't fully looked at the details of this PR, but based on the documentation example and this example, this "should" work correctly. The lambda for Layout.FromMethod would essentially look like the following:

` logEvent => new AzureServiceTokenProvider().GetAccessTokenAsync($"https://database.windows.net
/").Result;`

Technically `GetAccessTokenAsync` should be "awaited" and called from another async-capable function, but it appears from my looking that NLog does not have async piped through the entire stack at this point in time. The above approach will still work, and every time a connection is opened it would call `GetAccessTokenAsync` by rendering the layout, and get the latest token at the time it's called.

I'm happy to help as I am able are you are right, I initially was planning on providing an acceptable solution for this. My only concern, from a timing perspective, is I do not know the internals of NLog as thoroughly as you two do and it will take me a bit to come up to speed to where I feel comfortable not breaking anything. It sounds like you guys want to release 4.7.0 soon-ish.

File: src/NLog/Targets/DatabaseConnectionInfo.cs (hunk 7 above)
Comment (304NotModified) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: > Ok lets rename it to DatabasePropertyInfo.

👍 that's a good idea

Comment (snakefoot) [reply src/NLog/Targets/DatabaseConnectionInfo.cs]: Ooops it became DatabaseObjectPropertyInfo

**Comment by snakefoot** [comment]
@anxkha Would probably implement some fancy caching logic in your lambda, so you only refresh the AccessToken every 10-30 minutes (Maybe with an async-background-timer). So when calling the lambda it just returns the latest AccessToken.

**Comment by snakefoot** [comment]
Rename also required an extra push

**Comment by anxkha** [comment]
> Would probably implement some fancy caching logic in your lambda, so you only refresh the AccessToken every 10-30 minutes (Maybe with an async-background-timer). So when calling the lambda it just returns the latest AccessToken. Because the lambda will be called frequently (for every logevent unless doing batching with AsyncWrapper).

The `GetAccessTokenAsync` call actually does this for you. It will re-use the same token until it approaches expiration, then reach out to renew it for you so that you don't have to worry about it. I thought the same thing too, initially, but finally found a little more documentation on it.

**Comment by snakefoot** [comment]
> The GetAccessTokenAsync call actually does this for you. It will re-use the same token until it approaches expiration,

Again I know nothing about AccessToken, but it looks like it can perform a http-request to lookup the next token. I'm guessing Microsoft have added logging incase AccessToken lookup fails. It is strictly forbidden to perform logging while inside a NLog-target (Causes deadlocks or stackoverflows, both issues you don't want).

So I think it is very important that the custom `accessTokenLayout` is not the one doing the actual http-request but just returns the latest AccessToken-Value, and then have a background-time responsible for refreshing the AccessToken-Value every 10 min.

**Comment by snakefoot** [comment]
@anxkha Please tell if this custom pre-release solves you requirement. It will not enter the NLog 4.7 release, unless it actually meets the goal. It can take awhile before the next release train leaves the station.

**Comment by anxkha** [comment]
@snakefoot Yeah I'll do my best to test it today as things are calmer. Regarding setting the access token, I hear you. I don't know what goes on inside their library to do it other than one or two HTTP calls to Azure. I'll come up with some process to refresh it outside the cycle.

**Comment by anxkha** [comment]
@snakefoot I was able to test this and it appears to be working flawlessly! This is with build 4.6.0-11293.

**Comment by snakefoot** [comment]
@anxkha Wow that was quick, and good news. Would it be possible for you to share how you implemented the AccessToken-Lookup logic with background-refresh?

**Comment by anxkha** [comment]
@snakefoot Yep! In my first draft pass I have a `DatabaseConnectionFactory` class that has the following:

	public sealed class DatabaseConnectionFactory
	{
		private static readonly Timer _accessTokenTimer;
		...
		public static string AccessToken { get; private set; }

		static DatabaseConnectionFactory()
		{
			if (!UseMsiConnection)
				return;

			_accessTokenTimer = new Timer(60 * 1000)
			{
				AutoReset = true,
				Enabled = true
			};
			_accessTokenTimer.Elapsed += async (e, sender)
				=> AccessToken = await new AzureServiceTokenProvider().GetAccessTokenAsync($"https://{DatabaseHostSuffix}/");
			_accessTokenTimer.Start();

			AccessToken = new AzureServiceTokenProvider().GetAccessTokenAsync($"https://{DatabaseHostSuffix}/").Result;
		}
		...

And then for the NLog configuration:

	LogManager.LoadConfiguration("nlog.config");

	var databaseTarget = (DatabaseTarget)LogManager.Configuration.FindTargetByName("database");
	databaseTarget.ConnectionString = DatabaseConnectionFactory.SomeDatabase;
	databaseTarget.CommandType = CommandType.StoredProcedure;

	if (DatabaseConnectionFactory.UseMsiConnection)
	{
		var accessTokenLayout = Layout.FromMethod(logEvent => DatabaseConnectionFactory.AccessToken);
		databaseTarget.ConnectionProperties.Add(new DatabaseObjectPropertyInfo
		{
			Name = nameof(SqlConnection.AccessToken),
			Layout = accessTokenLayout
		});
	}

	LogManager.ReconfigExistingLoggers();

I still need to do some clean code refactoring, and it needs some form of semaphore or locking to protect the AccessToken interaction, but it works and it supports both Managed Service Identity and regular SQL Auth access scenarios.

**Comment by snakefoot** [comment]
@anxkha If you have the time, then I would become very happy if you could can verify this nuget-package (NLog.Extensions.AzureAccessToken.2.0.0.nupkg) is working:

https://ci.appveyor.com/project/JDetmar/nlog-extensions-azurestorage/builds/31792097/artifacts

It can be used like this:

```
<extensions>
  <add assembly="NLog.Extensions.AzureAccessToken" /> 
</extensions>
<targets>
    <target type="database">
         <connectionProperty name="AccessToken" layout="${AzureAccessToken:ResourceName=${gdc:DatabaseHostSuffix}}"  />
    </target>
</targets>
```

And before logging then you should do this:

```c#
NLog.GlobalDiagnosticsContext.Set("DatabaseHostSuffix", $"https://{DatabaseHostSuffix}/");
```

**Comment by snakefoot** [comment]
@anxkha Nuget-package is now available:

https://www.nuget.org/packages/NLog.Extensions.AzureAccessToken/

**Comment by anxkha** [comment]
@snakefoot Thanks, I'll take a look today!

**Comment by snakefoot** [comment]
@anxkha Any success in your adventures with [NLog.Extensions.AzureAccessToken](https://www.nuget.org/packages/NLog.Extensions.AzureAccessToken/) ?

**Comment by anxkha** [comment]
@snakefoot My apologies, I am juggling 10 different projects and forgot to reply to you. Yes, this package using the instructions you put on the original feature request appears to be working great!

**Comment by snakefoot** [comment]
@anxkha Excellent. Thank you for the testing and the confirmation.
